content/blog/taksim-modes-makam.md

```diff
@@ -0,0 +1,664 @@
+
+Still, in many cases, it is useful to be able to specify and restrict the modes under which a predicate is used.
+This is especially useful when we are trying to accurately model existing type inferencing algorithms.
+This comes up in bidirectional typing for example, where we have two different typing judgments with different modes:
+
+- $\Gamma \vdash e \Rightarrow \tau$, the type *synthesis* or type *inference* judgment, where the type $\tau$ is unknown and is being synthesized from $e$
+- $\Gamma \vdash e \Leftarrow \tau$, the type *analysis* or type *checking* judgment, where the type $\tau$ is known and is being checked against $e$
+
+```makam-hidden
+expr : type.
+typ : type.
+
+annot : expr -> typ -> expr.
+eunit : expr.
+lambda : (expr -> expr) -> expr.
+app : expr -> expr -> expr.
+
+tunit : typ.
+arrow : typ -> typ -> typ.
+```
+
+Modeling these judgments as predicates in λProlog/Makam, their types are the same:
+
+```makam-noeval
+type_synth : expr -> typ -> prop.
+type_check : expr -> typ -> prop.
+```
+
+However, the two predicates differ in terms of how they are supposed to be used. In both of them, the expression argument is supposed to be known, so it's an input parameter; in the analysis predicate, the type is an input parameter; whereas in the synthesis predicate, the type argument is an output parameter, since it is unknown and what the predicate will compute.
+
+To capture this intention of how predicates are supposed to be used, *mode declarations* are common in logic programming (and higher-order logic programming) languages; in that way, we can turn off the "generative" behavior of predicates for input parameters. These could look something like this, using informal syntax:
+
+```
+mode type_synth (input output).
+mode type_check (input input).
+```
+
+Makam does not support mode declarations though. This brings us to the main thing we'll explore in this post: can we add support for them, without changing the core language?
+
+Here's the structure of the rest of this.
+
+- What happens if we transcribe the bidirectional system for the STLC directly? [→ example, naively](#example-naively)
+- How can we model bidirectional typing accurately? [→ example, manually](#example-manually)
+- It would be nice to be able to fix this just by adding a mode declaration for `type_synth` and `type_check`. Can we add mode declarations to Makam? Is there a way to do this without changing the core language? [→ modal makam](#modal-makam)
+- Do simple mode declarations provide enough generality? (TODO) [→ modal makam, take 2](#todo)
+- Can we turn the bidirectional recipe into a Makam program, to "bidirectionalize" an existing typing declaration? (TODO) [→ writing recipes in makam](#todo)
+
+## example, naively
+
+```makam-hidden
+type_of_var : expr -> typ -> prop.
+```
+
+```makam-hidden
+%extend naive.
+type_synth : expr -> typ -> prop.
+type_check : expr -> typ -> prop.
+```
```

Comment:
In these kinds of "literate" posts one of the things I struggle with is finding a good balance in the code-to-prose ratio -- I want this to be relatively readable if one does not want to know the full details, but sometimes that means that missing details make things more confusing :smile:  One idea is to extract the Makam code into something readable for someone who wants to read the full code, adding comments etc.

In this case, which part was useful for understanding? Is it the declaration of `type_of_var`, or the fact that each section comes with its own namespace to avoid clashes between definitions?

Reply:
Yeah, this is indeed a tricky balance! I think it's more when I hit a new rule without a declaration I my brain needs to do work to recover the types of stuff that we're using to get my bearings. Kind of how I like it in papers when they put the form of judgement in a box above the rules sometimes. I wonder if some sort of 'code folding' could help - but it would be an involved change that I wouldn't want you to hold things up on. For example, [here is the Kotlin docs on defining variables](https://kotlinlang.org/docs/reference/basic-syntax.html#defining-variables):

- Folded:

  <img width="739" alt="Screen Shot 2020-01-15 at 1 52 20 pm" src="https://user-images.githubusercontent.com/695077/72400903-b50b5280-379e-11ea-99cb-a33f151bdee6.png">
- Unfolded:

  <img width="735" alt="Screen Shot 2020-01-15 at 1 52 28 pm" src="https://user-images.githubusercontent.com/695077/72400909-b9d00680-379e-11ea-88e9-ed87c1a56470.png">
- Markdown source:

  <img width="774" alt="Screen Shot 2020-01-15 at 1 47 32 pm" src="https://user-images.githubusercontent.com/695077/72400913-bdfc2400-379e-11ea-96b2-cc6603171c08.png">

Reply:
Thanks @brendanzab, that's a nice idea!
Another thing that could help for the issue you mention (it's useful to know the types of predicates when looking at their rules) is to be able to hover over a constant and get its type.